File: simulator/tests/m1/m1.py
import json
import logging

from simulator.wrappers.python.modeling.modeling import (
    Atomic,
    RunnableModelCfg,
    RunnableModelPortCfg,
    new_atomic_from_cfg,
)


class GeneratorIncremental(Atomic):
    def __init__(self, id: str, name: str, ports=None):
        super().__init__(id=id, name=name, ports=ports)
        self.value = 0
        self.color = ""
        self.storage = "base"

    # Initialize est appelée avant la simulation.
    def initialize(self) -> None:
        self.value = 0
        self.storage = "base"
        self.hold_in("active", 1.0)
        logging.info("Pute")

    # Exit est appelée après la simulation.
    def exit(self) -> None:
        # no-op pour l’instant
        pass

    # DeltInt : transition interne
    def delt_int(self) -> None:
        self.value += 1

        if self.value >= 3:
            self.passivate()
            self.storage = "gt 3"
        else:
            self.hold_in("active", 1.0)

    # DeltExt : transition externe
    # Ici, on ignore les inputs et on ajuste juste sigma.
    def delt_ext(self, e: float) -> None:
        self.continue_(e)

    # DeltCon : confluent (interne + externe en même temps).
    # On fait simple : interne prioritaire.
    def delt_con(self, e: float) -> None:
        self.delt_int()

    # Lambda : fonction de sortie
    # Envoie la valeur courante sur le port "out" sous forme JSON.
    def lambda_(self) -> None:
        try:
            out_port = self.get_port_by_name("out")
        except KeyError:
            # Si le modèle n'a pas de port "out" dans le manifest, on ne fait rien
            return

        payload = json.dumps({"value": self.value})
        out_port.add_value(payload)


def NewModel(config: dict) -> Atomic:
    """
    config = JSON de shared.RunnableModel envoyé par le runner.
    On en extrait juste ce qui nous intéresse pour créer l'Atomic Python.
    """
    # Si "ports" vaut None ou n'existe pas → on prend []
    raw_ports = config.get("ports") or []

    ports_cfg = [
        RunnableModelPortCfg(id=p["id"], type=p["type"])
        for p in raw_ports
    ]

    cfg = RunnableModelCfg(
        id=config["id"],
        name=config["name"],
        ports=ports_cfg,
    )

    return new_atomic_from_cfg(cfg, GeneratorIncremental)

File: simulator/tests/m1/m1.go
package main

import (
	"encoding/json"

	"devsforge/simulator/shared"
	modeling "devsforge/simulator/wrappers/go/modeling"
)

type GeneratorIncrementalParameters struct {
	value int
	color string
}

type GeneratorIncremental struct {
	modeling.Atomic
	Parameters GeneratorIncrementalParameters
	storage    string
}

func NewModel(cfg shared.RunnableModel) modeling.Atomic {
	base := &GeneratorIncremental{
		Atomic: modeling.NewAtomic(cfg),
		Parameters: GeneratorIncrementalParameters{
			value: 0,
			color: "",
		},
	}
	return base
}

// Initialize est appelée avant la simulation.
func (m *GeneratorIncremental) Initialize() {
	m.Parameters.value = 0
	m.storage = "base"
	m.HoldIn("active", 1.0)
}

// Exit est appelée après la simulation.
func (m *GeneratorIncremental) Exit() {
	// no-op pour l’instant
}

// DeltInt : transition interne
func (m *GeneratorIncremental) DeltInt() {
	m.Parameters.value++

	if m.Parameters.value >= 3 {
		m.Passivate()
		m.storage = "gt 3"
	} else {
		m.HoldIn("active", 1.0)
	}
}

// DeltExt : transition externe
// Ici, on ignore les inputs et on ajuste juste sigma.
func (m *GeneratorIncremental) DeltExt(e float64) {
	m.Continue(e)
}

// DeltCon : confluent (interne + externe en même temps).
// On fait simple : interne prioritaire.
func (m *GeneratorIncremental) DeltCon(e float64) {
	m.DeltInt()
}

// Lambda : fonction de sortie
// Envoie la valeur courante sur le port "out" sous forme JSON.
func (m *GeneratorIncremental) Lambda() {
	outPort, err := m.GetPortByName("out")
	if err != nil {
		return
	}

	payload, err := json.Marshal(map[string]int{
		"value": m.Parameters.value,
	})
	if err != nil {
		return
	}
	outPort.AddValue(payload)

}
